### ISSUE TITLE
WeakRefs and FinalizationGroup tests plan

### ISSUE THREAD

**Issue by leobalter** [tc39/test262#2339]
After much consideration and lengthy discussion with peers here at Bocoup and across other member orgs, I've decided to restore the tests for WeakRefs and FinalizationGroup with the following plan:

- [x] Revert the commit removing the tests using `$262.gc`.
- [x] Apply the Async GC functionality (#2329)
  - [x] and fix that file name and some other nit picks 
- [x] **Default to success** if the host can't collect the references.
- [x] We will add the metadata for Test262's maintainer usage.

**Comment by mhofman** [comment]
I'm wondering what the strategy should be for testing how an implementation handles the callback when the execution is triggered by the host, i.e. from [`HostCleanupFinalizationGroup`](https://tc39.es/proposal-weakrefs/#sec-host-cleanup-finalization-group).

I realize this invocation is optional for the host, but my concern is the behavior when the host does make the invocation. In particular, how a poisoned callback would be handled. Currently [d8 segfaults](https://bugs.chromium.org/p/v8/issues/detail?id=8179#c60), which is pretty bad.
Of course the real question is, should there be a testing mechanism to make sure the error is reported, since reporting is out of scope for JS engines alone and fall on the embedder side.

**Comment by leobalter** [comment]
> I'm wondering what the strategy should be for testing how an implementation handles the callback when the execution is triggered by the host, i.e. from HostCleanupFinalizationGroup.

AFAICT, there isn't any way to observe this. I don't think we could write a Test262 valid test without requiring many extra artifacts from the hosts. 

> Of course the real question is, should there be a testing mechanism to make sure the error is reported, since reporting is out of scope for JS engines alone and fall on the embedder side.

Test262 does not describe anything on how these are handled, this depends a lot on each test runner. e.g. Test262-harness will check for anything going to the stderr output.

**Comment by mhofman** [comment]
> > I'm wondering what the strategy should be for testing how an implementation handles the callback when the execution is triggered by the host, i.e. from HostCleanupFinalizationGroup.
> 
> AFAICT, there isn't any way to observe this. I don't think we could write a Test262 valid test without requiring many extra artifacts from the hosts.
> 

Isn't this the same as `ClearKeptObjects`? The host could provide a hook `runHostCleanupFinalizationGroupJobs()`, which returns a promise. When resolved, the cleanup callback should have been invoked.
In v8, this currently translates into a `setTimeout`.

> > Of course the real question is, should there be a testing mechanism to make sure the error is reported, since reporting is out of scope for JS engines alone and fall on the embedder side.
> 
> Test262 does not describe anything on how these are handled, this depends a lot on each test runner. e.g. Test262-harness will check for anything going to the stderr output.

I guess something printed to the stderr is an indicator of a failed test, so that makes sense. I assume that would also catch the segfault I experienced in d8.

**Comment by mhofman** [comment]
Btw, there is discussion in https://github.com/tc39/proposal-weakrefs/issues/157 to relax the atomicity of the execution step.

In that case I think we need to make clear the `asyncGc` hook should only complete successfully after all the execution steps have been performed, i.e. clearing all WeakRef targets and FinalizationGroup cells for objects that have been collected is not optional and must be done before resolving.

That's similar to [@leobalter's suggestion](https://github.com/tc39/test262/issues/2239#issuecomment-513336159) that the `gc()` hook should always trigger execution.

**Comment by leobalter** [comment]
in our case, the asyncGC hook receives all the references we expect to be cleared before completion, otherwise it says it can't remove the references and ends the test earlier (without a failure as requested). 

My suggestion on the more strict GC is blocked for points we already discussed. It's still my preferred option but it's up to me to require it.

I'll close this for now and we should create separate issues for any changes, this helps me a lot keeping a better track of new missing coverage.
